src/phoenix/experimental/evals/templates/template.py

```diff
@@ -45,49 +52,45 @@ def _get_delimiters(self, delimiters: List[str]) -> Tuple[str, str]:
+    def _validate(self, text: str) -> None:
         # Validate that for every open delimiter, we have the corresponding closing one
-        start_count = self.text.count(self._start_delim)
-        end_count = self.text.count(self._end_delim)
+        start_count = text.count(self._start_delim)
+        end_count = text.count(self._end_delim)
         if start_count != end_count:
             raise ValueError(
                 f"text poorly formatted. Found {start_count} instances of delimiter "
                 f"{self._start_delim} and {end_count} instances of {self._end_delim}. "
                 "They must be equal to be correctly paired."
             )
 
-    def _parse_variables(self) -> List[str]:
-        variables = []
-        formatter = Formatter()
-
-        text = self.text
-
-        # Example of this could be a template like: My name is ::name::
-        if self._start_delim == self._end_delim:
-            delim_length = len(self._start_delim)
-            delim_count = text.count(self._start_delim)
-            while delim_count > 0:
-                left_index = text.find(self._start_delim)
-                right_index = text[left_index + delim_length :].find(self._start_delim)
-                text = (
-                    text[0:left_index]
-                    + DEFAULT_START_DELIM
-                    + text[left_index + delim_length : left_index + delim_length + right_index]
-                    + DEFAULT_END_DELIM
-                    + text[left_index + 2 * delim_length + right_index :]
-                )
-                delim_count = text.count(self._start_delim)
-        else:
-            if self._start_delim != "{":
-                text = text.replace(self._start_delim, DEFAULT_START_DELIM)
-            if self._end_delim != "{":
-                text = text.replace(self._end_delim, DEFAULT_END_DELIM)
+    def _parse_variables(self, text: str) -> List[str]:
+        pattern = re.escape(self._start_delim) + "(.*?)" + re.escape(self._end_delim)
+        variables = re.findall(pattern, text)
+        return variables
 
-        for _, variable_name, _, _ in formatter.parse(text):
-            if variable_name:
-                variables.append(variable_name)
 
-        return variables
+class ClassificationTemplate(PromptTemplate):
+    def __init__(
+        self,
+        rails: Dict[bool, str],
+        base_template: str,
+        explanation_template: str,
+        delimiters: List[str] = [DEFAULT_START_DELIM, DEFAULT_END_DELIM],
+    ):
+        self.rails = rails
+        self.base_template = base_template
+        self.explanation_template = explanation_template
+        self._start_delim, self._end_delim = self._get_delimiters(delimiters)
+        self.variables: List[str] = []
+        for text in [base_template, explanation_template]:
+            self._validate(text)
+            self.variables += self._parse_variables(text)
+
+    def prompt(self, **options: Any) -> str:
```

Comment:
suggestion: can we strongly type the options here? It seems unnecessarily ambiguous and would prefer the least amount of `Any` usage as possible

Reply:
yep absolutely